### PR TITLE
upstream: locality weighted maglev load balancing.

### LINKED ISSUE #2725: Weighted load balancing
We have a need for weighted LB that respects the (v2) locality weighting in addition to endpoint weighting. This needs to work initially for RR and Maglev LB, but eventually we would like any implementation to generalize to all LBs (at least at the locality level). This issue will track design discussion and implementation work. 

We can weight RR via replication of hosts in the queue, at the expense of some memory use. Similarly, while populating the Maglev table, a similar approach can be taken.

One fundamental design question is whether we have true two-level hierarchical weighted LB (first pick locality, then pick endpoint within locality) or flat LB (combined locality weight and intra-locality endpoint weight for a single weight). Flat LB is conceptually simple, but may be expensive in terms of memory cost when we have a wide range of LB weights that combined weighting would imply. By having true hierarchical LB, we also get "for free" locality weighted LB for LBs that don't support weighting. The downside is we will need to implement at least a simple weighted locality LB and a consistent hash locality LB to support affinity and non-affinity second-level LBs.

I think based on the above it makes sense to do hierarchical LB. If this makes sense, I can work on weighted RR as a first step.

@mattklein123 @zuercher @alyssawilk WDYT?

### PR DESCRIPTION
**Pull request by htuch**
This follows up from #2982 and adds the ability to weight at the locality level when performing
Maglev LB. Unlike the non-affinity case, we don't do a two level LB (pick locality, pick host), but
use a single Maglev table. This reduces memory cost when there are few localities relative to hosts,
at the expense of reduced granularity. Ring hash LB is explicitly not supported for weighted,
locality or otherwise.

Fixes #2725.

Risk Level: Low.
Testing: Additional unit tests for locality weighted in maglev_lb_test.

Signed-off-by: Harvey Tuch <htuch@google.com>

### CODE REVIEW

File: source/common/upstream/maglev_lb.cc [hunk 1]
```diff
@@ -3,38 +3,60 @@
...
 
   // Compute maximum host weight. If this is zero, we are doing unweighted Maglev.
   uint32_t max_host_weight = 0;
-  for (const auto& host : hosts) {
-    max_host_weight = std::max(host->weight(), max_host_weight);
+  uint32_t total_hosts = 0;
+  for (uint32_t i = 0; i < hosts_per_locality.get().size(); ++i) {
+    for (const auto& host : hosts_per_locality.get()[i]) {
+      max_host_weight =
+          effective_weight(std::max(effective_weight(host->weight(), i), max_host_weight), i);
```
Comment (mattklein123) [inline source/common/upstream/maglev_lb.cc]: Having a hard time following the math here. Can you add a comment?

File: source/common/upstream/maglev_lb.h [hunk 2]
```diff
@@ -54,8 +58,29 @@ class MaglevLoadBalancer : public ThreadAwareLoadBalancerBase {
...
   // ThreadAwareLoadBalancerBase
-  HashingLoadBalancerSharedPtr createLoadBalancer(const HostVector& hosts) override {
-    return std::make_shared<MaglevTable>(hosts, table_size_);
+  HashingLoadBalancerSharedPtr createLoadBalancer(const HostSet& host_set) override {
+    // Note that we only compute global panic on host set refresh. Given that the runtime setting
+    // will rarely change, this is a reasonable compromise to avoid creating extra LBs when we only
+    // need to create one per priority level.
+    const bool has_locality =
+        host_set.localityWeights() != nullptr && !host_set.localityWeights()->empty();
+    if (isGlobalPanic(host_set)) {
```
Comment (mattklein123) [inline source/common/upstream/maglev_lb.h]: In looking at the code duplication here I wonder if there is a way to still have the panic computation happen generally, and then we just get passed either healthy or not healthy host sets, even if Ketama doesn't operate on a per locality basis? Might not be worth it but just throwing it out there. Mainly just wondering the reasoning behind deciding to duplicate vs. generalize.

Either way please check to make sure we have coverage on all these paths.

File: source/common/upstream/maglev_lb.cc (hunk 1 above)
Comment (htuch) [reply source/common/upstream/maglev_lb.cc]: That's because there's some typo there, will fix.

File: source/common/upstream/maglev_lb.h (hunk 2 above)
Comment (htuch) [reply source/common/upstream/maglev_lb.h]: Host sets include both healthy/unhealthy, it's only individual host lists or hosts per locality that can be split out like that. The rationale for the current structure is that I didn't want to complicate Ketama with dealing with localities.

We have full coverage, see https://37894-65214191-gh.circle-artifacts.com/0/build/envoy/generated/coverage/coverage.bazel-out_k8-dbg_bin_source_common_upstream__virtual_includes_maglev_lb_lib_common_upstream_maglev_lb.h.html and https://37894-65214191-gh.circle-artifacts.com/0/build/envoy/generated/coverage/coverage.bazel-out_k8-dbg_bin_source_common_upstream__virtual_includes_ring_hash_lb_lib_common_upstream_ring_hash_lb.h.html.

Comment (mattklein123) [reply source/common/upstream/maglev_lb.h]: OK fair enough. I guess we can revisit in the future if we find we keep duplicating this logic to more load balancers.

**Pull request merged**